recipe-server/client/control_new/components/extensions/ExtensionForm.js

```diff
@@ -0,0 +1,140 @@
+import { Button, Form, Icon, Input, Upload } from 'antd';
+import autobind from 'autobind-decorator';
+import { Map } from 'immutable';
+import React, { PropTypes as pt } from 'react';
+import { isEqual } from 'underscore';
+
+import BaseAntForm from 'control_new/components/common/BaseAntForm';
+import FormActions from 'control_new/components/common/FormActions';
+
+
+/**
+ * Form for editing XPIs uploaded to the service.
+ *
+ * The XPI field does not use normal Ant field wrappers due to how the Upload
+ * component works. Instead, the resulting file is stored in the form state and
+ * appended to the form values after validation.
+ */
+@autobind
+export class _ExtensionForm extends BaseAntForm {
+  static propTypes = {
+    extension: pt.instanceOf(Map),
+    form: pt.object,
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      xpiName: null,
+      xpiUrl: null,
+      xpiFile: null,
+    };
+  }
+
+  componentWillMount() {
+    const { extension } = this.props;
+    if (extension) {
+      this.setInitialValues(extension);
+    }
+  }
+
+  componentWillReceiveProps(newProps) {
+    // Reset form if the extension changed
+    if (!isEqual(newProps.extension, this.props.extension)) {
```

Comment:
I ended up using [`Immutable.is`](https://facebook.github.io/immutable-js/docs/#/is) to avoid trying to handle when `newProps.extension` is undefined.